sql/core/src/main/scala/org/apache/spark/sql/execution/window/RankLimitExec.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.apache.spark.sql.execution.window
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.codegen.LazilyGeneratedOrdering
+import org.apache.spark.sql.catalyst.plans.physical._
+import org.apache.spark.sql.execution.{GroupedIterator, SparkPlan, UnaryExecNode}
+import org.apache.spark.util.collection.Utils
+
+/**
+ * This operator is designed to filter out unnecessary rows before WindowExec,
+ * for top-k computation.
+ * @param partitionSpec Should be the same as [[WindowExec#partitionSpec]]
+ * @param orderSpec Should be the same as [[WindowExec#orderSpec]]
+ */
+case class RankLimitExec(
+    partitionSpec: Seq[Expression],
+    orderSpec: Seq[SortOrder],
+    limit: Int,
+    child: SparkPlan) extends UnaryExecNode {
+  assert(orderSpec.nonEmpty && limit > 0)
+
+  // apply Utils.takeOrdered when partitionSpec is empty and row_number is used.
+  private def applyTakeOrdered: Boolean = partitionSpec.isEmpty
+
+  override def output: Seq[Attribute] = child.output
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] = {
+    if (applyTakeOrdered) {
+      super.requiredChildOrdering
+    } else {
+      // Should be the same as [[WindowExec#requiredChildOrdering]]
+      Seq(partitionSpec.map(SortOrder(_, Ascending)) ++ orderSpec)
+    }
+  }
+
+  override def outputOrdering: Seq[SortOrder] = {
+    if (applyTakeOrdered) {
+      orderSpec
+    } else {
+      child.outputOrdering
+    }
+  }
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  // TODO: support rank and dense_rank
```

Comment:
How can we support `rank` and `dense_rank`?

Reply:
1, for `rank` and `dense_rank`, no matter whether partitionSpec is empty,  `requiredChildOrdering = Seq(partitionSpec.map(SortOrder(_, Ascending)) ++ orderSpec)`;
2, then in the map side, we can compute the `rank` and `dense_rank` for each row, and filter out unnecessary rows.

Reply:
For example:
```
0: jdbc:hive2://10.211.174.27:10000/access_vi> SELECT a, b, rank() OVER (PARTITION BY a ORDER BY b) as rk FROM VALUES ('A1', 1), ('A1', 1), ('A2', 3), ('A1', 1) tab(a, b);
+-----+----+-----+--+
|  a  | b  | rk  |
+-----+----+-----+--+
| A1  | 1  | 1   |
| A1  | 1  | 1   |
| A1  | 1  | 1   |
| A2  | 3  | 1   |
+-----+----+-----+--+
4 rows selected (8.259 seconds)
```
If we add a filter condition: `rk = 1`. How can we ensure the result is correct if we use `RankLimitExec`?

Reply:
within each maptask:
1, sort the partion by `(a,b)`;
2, group the iterator by `a`:  `GroupedIterator.apply(stream, partitionSpec, output)`
3, within each group, group again by `b`, compute the `rk`, and skip if `rk` > 1.

the `limitFunc` for `rank` and `dense_rank` may be like:

```
      case _: Rank =>
        (partitionedStream: Iterator[InternalRow]) =>
          new Iterator[InternalRow]() {
            private val internalIter =
              GroupedIterator.apply(partitionedStream, orderSpec.map(_.child), output)

            private var cnt = 0
            private var rank = 0

            private var currentGroup: Iterator[InternalRow] = null

            fetchNextGroup()

            private def fetchNextGroup(): Unit = {
              if (internalIter.hasNext && rank < limit) {
                currentGroup = internalIter.next()._2
                rank = cnt + 1
              } else {
                currentGroup = null
              }
            }

            override def hasNext: Boolean = currentGroup != null && currentGroup.hasNext

            override def next(): InternalRow = {
              val row = currentGroup.next()
              if (currentGroup.isEmpty) {
                fetchNextGroup()
              }
              cnt += 1
              row
            }
          }

      case _: DenseRank =>
        (partitionedStream: Iterator[InternalRow]) =>
          GroupedIterator
            .apply(partitionedStream, orderSpec.map(_.child), output)
            .take(limit)
            .flatMap(_._2)
```

Reply:
The correctness exists in: 
for all three rank functions (row_number, rank, dense_rank), `global rank` <= `partial rank`, so we can safely remove rows with `partial rank > limit`